ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java

```diff
@@ -368,7 +370,14 @@ void replicate(List<List<RemoteReplicaInfo>> replicasToReplicate) {
                   new ExchangeMetadataResponse(missingStoreKeys, replicaMetadataResponseInfo.getFindToken(),
                       replicaMetadataResponseInfo.getRemoteReplicaLagInBytes());
               exchangeMetadataResponseList.add(exchangeMetadataResponse);
+              storageManager.updateIOErrorCountPerStore(remoteReplicaInfo.getLocalReplicaId().getPartitionId(), false);
```

Comment:
Let's chat about this. I don't think we should use the `StorageManager` here for two reasons
1) New code that uses the `Store` has to remember to update these counts to keep the feature working correctly
2) Any new types of errors that require automatic shutdown will have to be counted (more code addition)

For those reasons, I think that all this error detection should be completely encapsulated with no external component having to explicitly tell the `StorageManager` that something happened.

One way to do this is to have all code in the functions in `BlobStore` under a `try..catch` which counts the relevant exceptions and issues a callback to `StorageManager` if there are problems.

Reply:
The code I'm suggesting is similar to what you have here but just the caller of this function should be `BlobStore` itself and not the components that use the `Store`

Reply:
Right. Let me check all the places to see if exception/error could be captured inside `BlobStore`. When I was writing this code, I also felt we shouldn't make `StorageManager` and `ReplicationManager` couple to tightly. I'll take your advice and make relevant changes.